File: tools/misc/git-patches/git-patches-manager.ts
import { brandingBaseName, brandingName } from "../../utils/defines.ts";
import { join } from "@std/path";
import {
  ensureDir,
  isExistsSync,
  isGitInitialized,
  runCommandChecked,
} from "../../utils/utils.ts";

// Helper to execute git commands
async function runGitCommand(
  args: string[],
  options: { cwd?: string } = {},
): Promise<{ stdout: string; stderr: string }> {
  return await runCommandChecked("git", args, options);
}

function getBinDir(): string {
  return Deno.build.os !== "darwin"
    ? `_dist/bin/${brandingBaseName}`
    : `_dist/bin/${brandingBaseName}/${brandingName}.app/Contents/Resources`;
}

const PATCHES_DIR = "tools/build/tasks/git-patches/patches";
const PATCHES_TMP = "_dist/bin/applied_patches";

export async function initializeBinGit() {
  const BIN_DIR = getBinDir();
  if (await isGitInitialized(BIN_DIR)) {
    console.log(
      "[git-patches] Git repository is already initialized in _dist/bin.",
    );
    return;
  }
  console.log("[git-patches] Initializing Git repository in _dist/bin.");
  // Create .gitignore
  await ensureDir(BIN_DIR);
  await Deno.writeTextFile(
    join(BIN_DIR, ".gitignore"),
    [
      `./noraneko-dev/*`,
      `./browser/chrome/browser/res/activity-stream/data/content/abouthomecache/*`,
    ].join("\n"),
  );
  // Initialize Git
  await runGitCommand(["init"], { cwd: BIN_DIR });
  await runGitCommand(["add", "."], { cwd: BIN_DIR });
  await runGitCommand(["commit", "-m", "initialize"], { cwd: BIN_DIR });
  console.log("[git-patches] Git repository initialization complete.");
}

export function checkPatchIsNeeded(): boolean {
  // Check if patch directory exists
  if (!isExistsSync(PATCHES_DIR, { isDirectory: true })) return false;
  if (!isExistsSync(PATCHES_TMP, { isDirectory: true })) return true;
  // Additional patch application logic can be added here
  return false;
}

// Additional patch management functions (applyPatches, createPatches) should be migrated here as needed.

File: tools/misc/git-patches/git-patches-manager.rb
require 'fileutils'
require 'open3'

BRANDING_BASE_NAME = 'noraneko'
BRANDING_NAME = 'Noraneko'

def bin_dir
  if /darwin/ =~ RUBY_PLATFORM
    "_dist/bin/#{BRANDING_BASE_NAME}/#{BRANDING_NAME}.app/Contents/Resources"
  else
    "_dist/bin/#{BRANDING_BASE_NAME}"
  end
end

PATCHES_DIR = 'tools/build/tasks/git-patches/patches'
PATCHES_TMP = '_dist/bin/applied_patches'

def git_initialized?(dir)
  File.exist?(File.join(dir, '.git'))
end

def initialize_bin_git
  dir = bin_dir
  if git_initialized?(dir)
    puts '[git-patches] Git repository is already initialized in _dist/bin.'
    return
  end
  puts '[git-patches] Initializing Git repository in _dist/bin.'
  FileUtils.mkdir_p(dir)
  File.write(File.join(dir, '.gitignore'), [
    './noraneko-dev/*',
    './browser/chrome/browser/res/activity-stream/data/content/abouthomecache/*'
  ].join("\n"))
  %w[init add . commit -m initialize].each_slice(2) do |cmd, arg|
    Open3.capture3('git', cmd, arg, chdir: dir)
  end
  puts '[git-patches] Git repository initialization complete.'
end

def patch_needed?
  Dir.exist?(PATCHES_DIR) && !Dir.exist?(PATCHES_TMP)
end